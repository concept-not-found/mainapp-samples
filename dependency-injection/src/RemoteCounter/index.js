import {h, Component} from 'mainapp'

export default Component({
  inject (state, apiService) {
    return {
      apiService
    }
  },
  loading: true,
  count: NaN,
  async loadCount (state) {
    const {apiService} = state
    const remoteCount = await apiService.loadCount()
    return {
      loading: false,
      count: remoteCount
    }
  },
  down ({count}, value) {
    return {
      count: count - value
    }
  },
  up ({count}, value) {
    return {
      count: count + value
    }
  },
  view ({loading, count, down, up}) {
    if (loading) {
      return <p>Loading...</p>
    }
    return <div>
      <h1>{count}</h1>
      <button onclick={() => down(1)}>-</button>
      <button onclick={() => up(1)}>+</button>
    </div>
  }
})
