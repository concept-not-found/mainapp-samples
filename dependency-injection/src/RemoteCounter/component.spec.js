import RemoteCounter from './'

describe('remote counter', () => {
  const specification = RemoteCounter.specification

  it('starts in a loading state', () => {
    expect(specification.loading).toBe(true)
  })

  it('loads the count from the api using loadCount', () => {
    const apiService = {
      async loadCount () {
        return 42
      }
    }
    expect(specification.loadCount({apiService})).resolves.toEqual({
      loading: false,
      count: 42
    })
  })

  it('decrements by value given to down', () => {
    const state = {
      count: 42
    }
    const value = 40
    expect(specification.down(state, value)).toEqual({
      count: 2
    })
  })

  it('increments by value given to up', () => {
    const state = {
      count: 42
    }
    const value = 40
    expect(specification.up(state, value)).toEqual({
      count: 82
    })
  })

  it('renders loading while still loading', () => {
    const state = {
      loading: true
    }
    expect(specification.view(state)).toMatchSnapshot()
  })

  it('renders the count and -/+ buttons when loaded', () => {
    const state = {
      loading: false,
      count: 42
    }
    expect(specification.view(state)).toMatchSnapshot()
  })
})
