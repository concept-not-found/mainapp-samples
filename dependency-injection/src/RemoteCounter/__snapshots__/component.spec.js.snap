// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remote counter renders loading while still loading 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    "Loading...",
  ],
  "key": null,
  "name": "p",
}
`;

exports[`remote counter renders the count and -/+ buttons when loaded 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {},
      "children": Array [
        42,
      ],
      "key": null,
      "name": "h1",
    },
    Object {
      "attributes": Object {
        "onclick": [Function],
      },
      "children": Array [
        "-",
      ],
      "key": undefined,
      "name": "button",
    },
    Object {
      "attributes": Object {
        "onclick": [Function],
      },
      "children": Array [
        "+",
      ],
      "key": undefined,
      "name": "button",
    },
  ],
  "key": null,
  "name": "div",
}
`;
